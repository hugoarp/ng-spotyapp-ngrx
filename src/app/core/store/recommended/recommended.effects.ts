import { Injectable } from '@angular/core';
import { createEffect, ofType } from '@ngrx/effects';
import { Actions } from '@ngrx/effects';
import { SpotifyService } from 'src/app/core/services/spotify.service';
import { EMPTY, take } from 'rxjs';
import { map, mergeMap, catchError, switchMap, tap } from 'rxjs/operators';
import {
  loadedRecommendedTracks,
  loadRecommendedTracks,
} from './recommended.actions';

@Injectable()
export class RecommendedTracksInfoEffects {
  constructor(private spotify: SpotifyService, private actions$: Actions) {}

  loadRecommendedTracks$ = createEffect(() =>
    this.actions$.pipe(
      ofType(loadRecommendedTracks.type),
      mergeMap(() => this.spotify.getRecommendations()),
      mergeMap((trackInfo) => {
        const ids = trackInfo.map((track) => track.id);
        return this.spotify.checkUserFavorites(ids).pipe(
          map((favorites) => {
            return trackInfo.map((track, index) => {
              track.favorite = favorites[index];
              return track;
            });
          }),
          map((trackInfo) => ({
            type: loadedRecommendedTracks.type,
            trackInfo,
          })),
          catchError(() => EMPTY)
        );
      }),
      take(1)
    )
  );
}

// switchMap((recommendations) => {

// }),

// return [loadedRecommendedTracks({ trackInfo })];
