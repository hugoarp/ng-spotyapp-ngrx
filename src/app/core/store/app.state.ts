import { ActionReducerMap } from '@ngrx/store';
import { TrackInfoState } from '../interfaces/track';
import { UserInfoState } from '../interfaces/user';
import { favoritesTrackInfoReducer } from './favorites/favorites.reducers';
import { recommendedTrackInfoReducer } from './recommended/recommended.reducers';
import { userInfoReducer } from './user/user.reducers';

export interface AppState {
  userInfo: UserInfoState;
  recommendedTrackInfo: TrackInfoState;
  favoritesTrackInfo: TrackInfoState;
}

export const ROOT_REDUCERS: ActionReducerMap<AppState> = {
  userInfo: userInfoReducer,
  recommendedTrackInfo: recommendedTrackInfoReducer,
  favoritesTrackInfo: favoritesTrackInfoReducer,
};
