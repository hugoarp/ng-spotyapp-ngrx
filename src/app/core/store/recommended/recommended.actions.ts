import { createAction, props } from '@ngrx/store';
import { TrackInfo } from '../../interfaces/track';

export const loadRecommendedTracks = createAction(
  '[Recommended] Load recommended tracks'
);

export const loadedRecommendedTracks = createAction(
  '[Recommended] Load recommended tracks success',
  props<{ trackInfo: TrackInfo[] }>()
);

export const toggleFavoriteTrack = createAction(
  '[Recommended] Toggle favorite track',
  props<{ trackInfo: TrackInfo }>()
);
