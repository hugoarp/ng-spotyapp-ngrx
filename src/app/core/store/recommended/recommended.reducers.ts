import { createReducer, on } from '@ngrx/store';
import { TrackInfoState } from '../../interfaces/track';
import {
  loadedRecommendedTracks,
  loadRecommendedTracks,
} from './recommended.actions';

export const initialRecommendedTrackInfoState: TrackInfoState = {
  loading: false,
  tracks: [],
};

export const recommendedTrackInfoReducer = createReducer(
  initialRecommendedTrackInfoState,
  on(loadRecommendedTracks, (state) => {
    return { ...state, loading: true };
  }),
  on(loadedRecommendedTracks, (state, { trackInfo }) => {
    return { ...state, loading: false, tracks: trackInfo };
  })
);
