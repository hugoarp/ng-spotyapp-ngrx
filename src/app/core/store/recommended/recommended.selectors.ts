import { createSelector } from '@ngrx/store';
import { TrackInfoState } from '../../interfaces/track';
import { AppState } from '../app.state';

export const selectRecommendedTracksFeature = (state: AppState) =>
  state.recommendedTrackInfo;

export const selectRecommendedTracks = createSelector(
  selectRecommendedTracksFeature,
  (state: TrackInfoState) => state.tracks
);

export const selectRecommendedTracksLoading = createSelector(
  selectRecommendedTracksFeature,
  (state: TrackInfoState) => state.loading
);
