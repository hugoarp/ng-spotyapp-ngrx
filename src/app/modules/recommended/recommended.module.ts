import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';

import { RecommendedRoutingModule } from './recommended-routing.module';
import { RecommendedComponent } from './page/recommended.component';
import { TemplatesModule } from 'src/app/shared/components/templates/templates.module';

@NgModule({
  declarations: [RecommendedComponent],
  imports: [CommonModule, RecommendedRoutingModule, TemplatesModule],
})
export class RecommendedModule {}
