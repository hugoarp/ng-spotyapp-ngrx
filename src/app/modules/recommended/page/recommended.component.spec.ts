import { ComponentFixture, TestBed } from '@angular/core/testing';
import { provideMockStore } from '@ngrx/store/testing';

import { RecommendedComponent } from './recommended.component';

describe('RecommendedComponent', () => {
  let component: RecommendedComponent;
  let fixture: ComponentFixture<RecommendedComponent>;
  let compiled: HTMLElement;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [RecommendedComponent],
      providers: [provideMockStore({})],
    }).compileComponents();

    fixture = TestBed.createComponent(RecommendedComponent);
    component = fixture.componentInstance;
    fixture.detectChanges();
    compiled = fixture.nativeElement;
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  test('debe hacer match con el snapshot', () => {
    expect(compiled).toMatchSnapshot();
  });
});
