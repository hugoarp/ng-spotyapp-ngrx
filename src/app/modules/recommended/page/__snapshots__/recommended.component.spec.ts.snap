// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecommendedComponent debe hacer match con el snapshot 1`] = `
<div
  id="root1"
>
  
</div>
`;
